LibOS/shim/src/fs/proc/attestation.c

```diff
@@ -0,0 +1,399 @@
+#include <asm/fcntl.h>
+#include <asm/mman.h>
+#include <asm/prctl.h>
+#include <asm/unistd.h>
+#include <errno.h>
+#include <linux/fcntl.h>
+#include <pal.h>
+#include <pal_error.h>
+#include <shim_fs.h>
+#include <shim_internal.h>
+
+#include "sgx_api.h"
+#include "sgx_attest.h"
+
+// TODO: For some reason S_IF* macros are missing if this file is included before our headers. We
+// should investigate and fix this behavior.
+#include <linux/stat.h>
+
+static int proc_sgx_attestation_mode(const char* name, mode_t* mode) {
+    // The path is implicitly set by calling this function
+    __UNUSED(name);
+    *mode = 0444;
+    return 0;
+}
+
+static int proc_sgx_attestation_stat(const char* name, struct stat* buf) {
+    // The path is implicitly set by calling this function
+    __UNUSED(name);
+    memset(buf, 0, sizeof(struct stat));
+    buf->st_dev = buf->st_ino = 1;
+    buf->st_mode              = 0444 | S_IFREG;
+    buf->st_uid               = 0;
+    buf->st_gid               = 0;
+    buf->st_size              = 0;
+    return 0;
+}
+
+static __sgx_mem_aligned sgx_report_t report;
+static __sgx_mem_aligned sgx_target_info_t target_info;
+static __sgx_mem_aligned sgx_target_info_t my_target_info;
+static __sgx_mem_aligned sgx_report_data_t report_data;
+
+/* IAS interactions are cached and only repeated if this variable is false. */
+static int    ias_valid = 0;
+static char   ias_report[10 * 1024];
+static char   ias_header[10 * 1024];
+static size_t ias_report_size = sizeof(ias_report);
+static size_t ias_header_size = sizeof(ias_header);
+static int    quote_valid = 0;
+static uint8_t quote[2048];
+static size_t quote_size = sizeof(quote);
+
+static int proc_sgx_report_open(struct shim_handle* hdl, const char* name, int flags) {
+    __UNUSED(name);
+
+    if (flags & (O_WRONLY | O_RDWR))
+        return -EACCES;
+
+    struct shim_str_data* data = calloc(1, sizeof(struct shim_str_data));
+    if (!data) {
+        /* free(str); */
+        return -ENOMEM;
+    }
+
+    sgx_report(&target_info, &report_data, &report);
+
+    data->str          = (char*) &report;
+    data->len          = sizeof(sgx_report_t);
+    data->do_not_free  = 1;
+
+    hdl->type          = TYPE_STR;
+    hdl->flags         = flags & ~O_RDONLY;
+    hdl->acc_mode      = MAY_READ;
+    hdl->info.str.data = data;
+    hdl->info.str.ptr  = (char*) &report;
+
+    return 0;
+}
+
+static int proc_sgx_report_mode(const char* name, mode_t* mode) {
+    __UNUSED(name);
+    *mode = 0444;
+    return 0;
+}
+
+static int proc_sgx_ias_header_open(struct shim_handle* hdl, const char* name, int flags) {
```

Comment:
Dont like the idea of having 2 separate pseudo interfaces, one for ias_header and one for ias_report, given that we get both in one call to DkIASReport.  There could be discrepancy, if caller has a wrong sequence, for example, calls ias_header, updates report_data, and then fetches ias_report.